src/building/bootstrapping.md

```diff
@@ -4,63 +4,101 @@ This subchapter is about the bootstrapping process.
-  and its 'target' or 'object' libraries (`std` and `rustc`).
-  Both are staged, but in a staggered manner.
-- **Stage 1:** the code in your clone (for new version) is then
-  compiled with the stage0 compiler to produce the stage1 compiler.
-  However, it was built with an older compiler (stage0), so to
-  optimize the stage1 compiler we go to next the stage.
-  - In theory, the stage1 compiler is functionally identical to the
-    stage2 compiler, but in practice there are subtle differences. In
-    particular, the stage1 compiler itself was built by stage0 and
-    hence not by the source in your working directory: this means that
-    the symbol names used in the compiler source may not match the
-    symbol names that would have been made by the stage1 compiler. This is
-    important when using dynamic linking and the lack of ABI compatibility
-    between versions. This primarily manifests when tests try to link with any
-    of the `rustc_*` crates or use the (now deprecated) plugin infrastructure.
-    These tests are marked with `ignore-stage1`.
-- **Stage 2:** we rebuild our stage1 compiler with itself to produce
-  the stage2 compiler (i.e. it builds itself) to have all the _latest
-  optimizations_. (By default, we copy the stage1 libraries for use by
-  the stage2 compiler, since they ought to be identical.)
-- _(Optional)_ **Stage 3**: to sanity check our new compiler, we
-  can build the libraries with the stage2 compiler. The result ought
-  to be identical to before, unless something has broken.
-
-The `stage2` compiler is the one distributed with `rustup` and all other
-install methods. However, it takes a very long time to build because one must
-first build the new compiler with an older compiler and then use that to
-build the new compiler with itself. For development, you usually only want
-the `stage1` compiler: `x.py build library/std`.
+[Bootstrapping] is the process of using a compiler to compiler a later version of itself.
+
+This raises a [chicken-or-the-egg paradox](https://en.wikipedia.org/wiki/Chicken_or_the_egg): what rust compiler was used to compile the very first rust compiler?
+The answer is that the first compiler was not written in rust. It was
+[written in OCaml][ocaml-compiler]. Of course, it has long been discarded and since then the
+only compiler that is able to compile some version of `rustc` is a slightly earlier version of `rustc`.
+
+For this purpose a python script `x.py` is provided at the root of the repository.
+`x.py` downloads a pre-compiled compiler—the stage 0 compiler—and with it compiles from the current source code a compiler—the stage 1 compiler.
+Additionaly, it may use the stage 1 compiler to compile from the current source code another compiler—the stage 2 compiler.
+Below describes this process in some detail, including the reason for a stage 2 compiler and more.
+
+## The stages of bootstrapping
+
+Each stage involves:
+<!-- What is this set of dependencies? -->
```

Comment:
https://rustc-dev-guide.rust-lang.org/building/bootstrapping.html#what-is-a-sysroot